File: models/pos_order_line.py
from odoo import api, fields, models

class PosOrderLine(models.Model):
    _inherit = 'pos.order.line'

    from_location = fields.Many2one("stock.location", string="From Location")

    #add field to load into pos frontend
    @api.model
    def _load_pos_data_fields(self, config_id):
        fields = super(PosOrderLine, self)._load_pos_data_fields(config_id)
        return fields + ['from_location']

File: static/src/overrides/pos/pos_order_line.js
/** @odoo-module */

import {patch} from "@web/core/utils/patch";
import {PosOrderline} from "@point_of_sale/app/models/pos_order_line";

patch(PosOrderline, {
    extraFields: {
        ...(PosOrderline.extraFields || {}),
        from_location: {
            model: "pos.order.line",
            name: "from_location",
            relation: "stock.location",
            type: "many2one",
            local: false,
        }
    }
})

patch(PosOrderline.prototype, {
    setup(vals) {
        super.setup(vals);
        if (this.product_id) {
            this.attachWarehouseLocation(vals);
        }
    },

    //don't merge the same product but with different locations
    can_be_merged_with(orderline) {
        let result = super.can_be_merged_with(orderline);
        return result && this.raw.from_location.id === orderline.from_location;
    },

    //cannot set this.from_location due to serialize() function of Base component
    attachWarehouseLocation(vals) {
        this.raw.from_location = {id: vals.from_location}
    },

    getLocationId() {
        return this.raw.from_location.id
    }
});
